node/network/protocol/src/peer_set.rs

```diff
@@ -0,0 +1,115 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! All peersets and protocols used for parachains.
+
+use sc_network::config::{NonDefaultSetConfig, SetConfig};
+use std::borrow::Cow;
+use strum::{IntoEnumIterator, EnumIter};
+
+/// The peer-sets and thus the protocols which are used for the network.
+#[derive(Debug, Clone, Copy, PartialEq, EnumIter)]
+pub enum PeerSet {
+	/// The validation peer-set is responsible for all messages related to candidate validation and communication among validators.
+	Validation,
+	/// The collation peer-set is used for validator<>collator communication.
+	Collation,
+	/// Direct peer connections for availability distribution (and recovery?).
+    AvailabilityDistribution,
+}
+
+/// Protocol name as understood in substrate.
+///
+/// Ideally this would be defined in substrate as a newtype.
+type ProtocolName = Cow<'static, str>;
+
+impl PeerSet {
+	/// Get `sc_network` peer set configurations for each peerset.
+	///
+	/// Those should be used in the network configuration to register the protocols with the
+	/// network service.
+	pub fn get_info(self) -> NonDefaultSetConfig {
+		let protocol = self.into_protocol_name();
+		match self {
+			PeerSet::Validation => NonDefaultSetConfig {
+				notifications_protocol: protocol,
+				set_config: sc_network::config::SetConfig {
+					in_peers: 25,
+					out_peers: 0,
+					reserved_nodes: Vec::new(),
+					non_reserved_mode: sc_network::config::NonReservedPeerMode::Accept,
+				},
+			},
+			PeerSet::Collation => NonDefaultSetConfig {
+				notifications_protocol: protocol,
+				set_config: SetConfig {
+					in_peers: 25,
+					out_peers: 0,
+					reserved_nodes: Vec::new(),
+					non_reserved_mode: sc_network::config::NonReservedPeerMode::Accept,
+				},
+			},
+            PeerSet::AvailabilityDistribution => NonDefaultSetConfig {
+                notifications_protocol: protocol,
+                set_config: sc_network::config::SetConfig {
+                    in_peers: 25,
+                    out_peers: 0,
+                    reserved_nodes: Vec::new(),
+                    non_reserved_mode: sc_network::config::NonReservedPeerMode::Accept,
+                },
+            },
+		}
+	}
+
+	/// Get the protocol name associated with each peer set as static str.
+	pub const fn get_protocol_name_static(self) -> &'static str {
+		match self {
+			PeerSet::Validation => "/polkadot/validation/1",
+			PeerSet::Collation => "/polkadot/collation/1",
+            PeerSet::AvailabilityDistribution => "/polkadot/availability/1",
+		}
+	}
+
+	/// Convert a peer set into a protocol name as understood by Substrate.
+	///
+	/// With `ProtocolName` being a proper newtype we could use the `Into` trait here.
+	pub fn into_protocol_name(self) -> ProtocolName {
```

Comment:
why not `impl From<PeerSet> for ProtocolName`?

Reply:
actually this works: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5142caf2118a6850a6d2896915f40ce3

https://github.com/Ixrec/rust-orphan-rules

Reply:
Definitely, but I do not think it is valuable to have it for the generic `Cow` type. In that case I think a special method like `into_protocol_name` would be preferable.

into() ... with type signature PeerSet -> `ProtocolName` is pretty self explanatory, not so much with PeerSet -> Cow<'static, str>, therefore I opted for a more descriptive name.